DOC-800 Edge how-to using content bundle (#1400)
spectrocloud/librarium: content/docs/04-clusters/03-edge/04-edgeforge-workflow/12-build-artifacts.md

<br />

- Software dependencies are pre-loaded into the installer image.

- Optimizes the deployment process for bandwidth-constrained environments or air-gapped environments.

- The ability to more granularly manage the software dependencies available to Edge clusters. 

This how-to guide provides instructions for creating and using a content bundle to build the Edge artifacts. You will begin with installing a necessary tool, the Palette Edge CLI, on your development machine. The Palette Edge CLI is a command-line utility to interact with Palette and perform specific tasks in your development environment, such as creating a content bundle. Next, you will download all the software dependencies mentioned in your cluster profile using the Palette Edge CLI and create a content bundle. Lastly, when your content bundle is ready, you will use the CanvOS utility to embed the content bundle and user data into the Edge installer ISO image.    

The diagram below displays the overarching steps to build the Edge installer ISO using a content bundle. The diagram also highlights the primary prerequisites to create a content bundle.

> ![An overarching diagram displaying the workflow in the current guide.](/clusters_edge_edge-forge-workflow_build-images_build-artifacts_overarching.png)

Reviewer: Great diagram @sudkul87

# Prerequisites

<WarningBox>

  This how-to guide extends the [Build Edge Artifacts](/clusters/edge/edgeforge-workflow/palette-canvos) workflow. Therefore, you must complete it before proceeding with the current guide. 

</WarningBox>
  

To complete this guide, you will need the following items:
<br />

* A physical or virtual Linux machine with *AMD64* (also known as *x86_64*) processor architecture to build the Edge installer ISO image. You can issue the following command in the terminal to check your processor architecture. 
  <br/>

  ```bash
  uname -m
  ```

* The Linux machine should have the following minimum hardware configuration:
  - 4 CPU
  - 8 GB memory
  - 100 GB storage. The actual storage will depend on the size of the content bundle you will use to build the Edge installer ISO image.

* You must have completed the [Build Edge Artifacts](/clusters/edge/edgeforge-workflow/palette-canvos) guide to build the provider images and create a cluster profile referencing one of the provider images. 

> * A Spectro Cloud API key. Later in this guide, you will use this API key to authenticate the Palette Edge CLI utility and allow it to interact with the Palette. Refer to the [User Authentication](https://docs.spectrocloud.com/user-management/user-authentication/#apikey) guide to create a new API key. 

Reviewer: ```suggestion
* A Spectro Cloud API key. Later in this guide, you will use this API key to authenticate the Palette Edge CLI utility and allow it to interact with Palette. Refer to the [User Authentication](https://docs.spectrocloud.com/user-management/user-authentication/#apikey) guide to create a new API key. 
```

# Instructions

Use the following instructions on your Linux machine, which this guide refers to as the development environment. 
<br />

> 1.  Visit the [Downloads](https://docs.spectrocloud.com/spectro-downloads#paletteedgecli) page and download the latest Palette Edge CLI. You can download the Palette Edge CLI by clicking on the available URL or using the download URL in the following command. Replace the `[PALETTE-EDGE-BINARY-URL]` placeholder with the download URL. 

Reviewer: I think it might be easier to tell the user to download using the browser, they can right-hand click on the URL or just click on it. Most browsers will start the download after clicking on the URL. 

If we do that, then we can remove the `curl` step. And go straight to step 2.

<br />

  ```bash
  curl [PALETTE-EDGE-BINARY-URL] --output palette-edge
  ```

2. Open a terminal session and navigate to the folder where you have downloaded the palette-edge binary.
 
 

> 3. Set the executable permissions for the palette-edge binary by issuing the following command.

Reviewer: We can combine these two.

  | OS Flavor     | Description        | Base Image URI                    |
  | opensuse-leap | Opensuse Leap 15.4 | quay.io/kairos/core-opensuse-leap |
  | ubuntu-20     | Ubuntu 20.4 LTS    | quay.io/kairos/core-ubuntu-20-lts |
  | ubuntu-22     | Ubuntu 22.4 LTS    | quay.io/kairos/core-ubuntu-22-lts |
  └────────────────────────────────────────────────────────────────────────┘
  ┌─────────────────────────────────────────────────────────────────────────────────────────────┐
  | K8S Flavor | Description        | Supported Versions                                        |
  | k3s        | Rancher K3s        | 1.25.2-k3s1,1.24.6-k3s1,1.23.12-k3s1,1.22.15-k3s1         |
  | kubeadm    | Kubernetes kubeadm | 1.25.2,1.24.6,1.23.12,1.22.15                             |
  | rke2       | Rancher RK2        | 1.25.2-rke2r1,1.24.6-rke2r1,1.23.12-rke2r1,1.22.15-rke2r1 |
  └─────────────────────────────────────────────────────────────────────────────────────────────┘
  ┌─────────────────────────────────┐
  | Component             | Version |
  | Spectro Agent Version | v3.4.3  |
  | Kairos Version        | v2.0.3  |
  └─────────────────────────────────┘
  ```
<br />

> 6. Set the Spectro Cloud API key as an environment variable by issuing the following command. Replace the `[USE-YOUR-API-KEY_HERE]` placeholder with your API key. The Palette Edge CLI will use this API key to authenticate itself when interacting with the Spectro Cloud API. Once authenticated, the Palette Edge CLI can interact with your Palette account.

Reviewer: ```suggestion
6. Set the Spectro Cloud API key as an environment variable by issuing the following command. Replace the `[USE-YOUR-API-KEY_HERE]` placeholder with your API key. The Palette Edge CLI will use this API key to authenticate with Palette. Once authenticated, the Palette Edge CLI can interact with your Palette account.
```

<br />

  ```bash
  export API_KEY=[USE-YOUR-API-KEY_HERE]
  ```

7. Log in to [Palette](https://console.spectrocloud.com).

> 8. Copy the Palette project ID to use later in this guide. The project ID is on the top-right corner of your Palette project overview page. For example, the screenshot below highlights a project ID.

Reviewer: ```suggestion
8. Copy your Palette project ID.  You will use this ID in a later step. The project ID is on the top-right corner of your Palette project overview page. Use the following screenshot to help you find your project ID.
```
Minor language cleanup.

  ![A screenshot highlighting the project ID in Palette project overview page](/clusters_edge_edge-forge-workflow_build-images_build-project_id.png)

9. Navigate to the left **Main Menu** and select **Profiles**. 

> 10. Click on the specific cluster profile to view its details. 

Reviewer: ```suggestion
10. Select the cluster profile you want to include in the content bundle. Click on the target cluster profile to access its details page.
```
Add a bit more context to help the reader.

  ![A screenshot highlighting the cluster profile ID and project ID in the URL of the cluster details page.](/clusters_edge_edge-forge-workflow_build-images_build-artifacts_url.png)

12. Copy the cluster profile ID from the cluster details page URL for the next step. 

13. Switch back to your development environment, and set the project ID as an environment variable by issuing the following command. Replace the `[USE-YOUR-PROJECT-ID_HERE]` placeholder with your project ID.
<br />

  ```bash
  export PROJECT_ID=[USE-YOUR-PROJECT-ID_HERE]
  ```

14. Set the cluster profile ID as an environment variable using the following command. Replace the `[USE-YOUR-PROFILE-ID_HERE]` placeholder with your cluster profile ID. The Palette Edge CLI uses the cluster profile ID to reference the correct cluster profile and download all its software dependencies. 
<br />

  ```bash
  export PROFILE_ID=[USE-YOUR-PROFILE-ID_HERE]
  ```

> 15. Use the command below to create the content bundle. The command uses the following flags:

Reviewer: ```suggestion
15. Issue the command below to create the content bundle. The `build` command uses the following flags:
```

  |**Command Flag**|**Value**|
  |---|---|
  |`--api-key`|Spectro Cloud API key|
  |`--project-id`|Palette project ID|
  |`--cluster-profile-ids`|Cluster profile IDs. If you want to include multiple cluster profiles in the content bundle, add multiple cluster profile IDs separated by a comma.|
  |`--palette-endpoint`|Palette API endpoint. The default Palette API endpoint is `api.spectrocloud.com`|
  |`--outfile`|Path to write the final content bundle. |
  You can issue `palette-edge build --help` to know about other available flags. 
<br />

  ```bash
  palette-edge build --api-key $API_KEY \
    --project-id $PROJECT_ID \
    --cluster-profile-ids $PROFILE_ID \
    --palette-endpoint api.spectrocloud.com \
    --outfile content
  ```
  

16. Use the command below to list all files in the current directory to verify that you created the content bundle successfully. The content bundle will have the following naming convention, `content-[randon-string]`, for example, **content-8e61a9e5**. 
<br />

  ```bash
  ls -al
  ```

> 17. List the files in the content bundle folder using the following command. Replace the `content-[randon-string]` with the content bundle in your current directory. The output will display the compressed files for the core and the app content. 

Reviewer: ```suggestion
17. List the files in the content bundle folder using the following command. The output will display the compressed core and app content files. 
```

<br />

  ```bash

>   ls -l content-[randon-string]

Reviewer: ```suggestion
  ls -la content-*/
```
We can use `ls -la content-*/` . This way, the user only sees folders that start with `content-` and their content.

  ```
  ```bash hideClipboard
  # Sample output   
  total 3981104
  -rw-rw-r-- 1 jb jb 1598552722 Jul 26 18:20 app-content-8e61a9e5.zst
  -rw-rw-r-- 1 jb jb 2478086360 Jul 26 18:20 core-content-8e61a9e5.zst
  ```

> 18. Issue the following command to execute the **earthly.sh** script with elevated privileges. The `+iso` option specifies the build target. This command will generate an ISO image from the content bundle and other configurations you have specified in the **.arg** and **user-data** files. 

Reviewer: ```suggestion
18. Issue the following command to build the Edge artifacts with your content bundle. The `+iso` option specifies the build target. This command will generate an ISO image from the content bundle and other configurations you have specified in the **.arg** and **user-data** files. 
```
We don't need to explain earthly, let's just stay high level and state this will build the Edge artifacts with their content bundle.

<br />

  ```bash
  sudo ./earthly.sh +iso
  ```
  This command may take up to 15-20 minutes to finish depending on the resources of the host machine. 

# Validate

List the Edge installer ISO and checksum by issuing the following command from the **CanvOS/** directory.
<br />

```shell
ls build/
```

```shell hideClipboard
# Output
palette-edge-installer.iso      
palette-edge-installer.iso.sha256
```
<br />

To validate, you can prepare an edge device using the Edge installer ISO by following the steps below:
<br />

1. Create a bootable USB flash drive using any third-party software. Most software that creates a bootable USB drive will validate the ISO image.

> 2. Select a physical or virtual host machine to emulate as an edge device. Enable (Dynamic Host Configuration Protocol) DHCP on the edge device before proceeding with the installation process. Enabling DHCP is necessary for the device to obtain an IP address automatically from the network. 

Reviewer: ```suggestion
2. Select a physical or virtual host machine to emulate as an edge device. Enable (Dynamic Host Configuration Protocol) DHCP on the host before proceeding with the installation process. Enabling DHCP is necessary for the device to obtain an IP address automatically from the network. 
```
I think host is the right word here in case they go with VMware to validate the installer.

3. Flash the edge device with a bootable USB drive. 

> Finally, you must deploy an Edge cluster to test the edge device you prepare using the Edge installer ISO. 

Reviewer: ```suggestion
The last step is to power the edge device and start the installation process. For more information, refer to the [Perform Site Install](/clusters/edge/site-deployment/site-installation) documentation.

<InfoBox>

We recommend you check out the end-to-end tutorial [Deploy an Edge Cluster on VMware](/clusters/edge/site-deployment/deploy-cluster) for a detailed walkthrough on deploying an Edge environment using VMware to simulate bare metal hosts. You can use your VMware environment to validate the Edge artifacts and that you have all the required dependencies included in your content bundle. 

</InfoBox>
```
We can help the reader out a bit more by adding the links to where they should review information next.  I was thinking the infobox but feel free to change whatever you don't like.

</WarningBox>
<br />

In this guide, you will begin with installing the necessary tool, the Palette Edge CLI, on your development machine. Next, you will create a cluster profile in Palette referencing one of the provider images you built using the [Build Edge Artifacts](/clusters/edge/edgeforge-workflow/palette-canvos) workflow. Then, you will access the cluster profile using the Palette Edge CLI utility to create a content bundle. Lastly, when your content bundle is ready, you will use the CanvOS utility to embed the content bundle, site-specific configuration, and user data into the Edge installer ISO image.  
  

The diagram below displays the high-level steps to build the Edge installer ISO image using a content bundle.

![An overarching diagram displaying the workflow in the current guide.](/clusters_edge_edge-forge-workflow_build-images_build-artifacts_overarching.png)

> # Prerequisites

Reviewer: Besides the resource prereqs (CPU, Mem, Storage). You can make the bullet below the first bullet and then simplify all the other bullets without mentioning the previous guide again. 

* To complete this guide, you must have completed the [Build Edge Artifacts](/clusters/edge/edgeforge-workflow/palette-canvos) guide and generated all the required Edge artifacts. The required Edge artifacts are an Edge Installer ISO, and provider images created and uploaded to an image registry.

To complete this how-to guide, you will need the following items:
<br />

* All prerequisites outlined in the [Build Edge Artifacts](/clusters/edge/edgeforge-workflow/palette-canvos) guide, except for this guide, your Linux machine must have 100 GB of storage or higher. The actual storage will depend on the size of the content bundle you will use to build the Edge installer ISO image.

* Provider images are built following the [Build Edge Artifacts](/clusters/edge/edgeforge-workflow/palette-canvos) guide.

* One of the provider images is pushed to an image registry. This guide uses a provider image compatible with K3s v1.25.x pushed to the [ttl.sh](https://ttl.sh/) image registry. However, you can use any other provider image and registry per your requirements.

# Instructions

Use the following instructions on your Linux machine, which this guide refers to as the development environment.
<br />

> 1. The first few steps will guide you to download, install, and verify the Palette Edge CLI. The Palette Edge CLI is a command-line utility to interact with Palette and perform specific tasks in your development environment, such as creating a content bundle. 
> 
>   Issue the following command to download the Palette Edge CLI. This command uses `curl` to download the Palette Edge CLI binary from the specified URL and save it as **palette-edge** in the current directory. 
> <br />

Reviewer: ```suggestion
1.  Visit the [Downloads](https://docs.spectrocloud.com/spectro-downloads#paletteedgecli) page and download the latest Palette Edge CLI. You can download the Palette Edge CLI by clicking on the available URL.  The Palette Edge CLI is a command-line utility to interact with Palette and perform specific tasks in your development environment, such as creating a content bundle. 

 2. Open up a terminal session and navigate to the **Downloads** folder.
 
 
 3. Set the executable permissions for the palette-edge binary by issuing the following command.

<br />
```

A couple of changes. The overview steps are redundant and not needed. The other thing that makes our life easier is by sending the users to our downloads page and telling them to download the latest version. That way you don't have to keep updating the curl command.

  ```bash
  curl https://software.spectrocloud.com/stylus/v3.4.3/cli/linux/palette-edge --output palette-edge
  ```

2. Set the executable permissions for the **palette-edge** binary by issuing the following command.
<br />

  ```bash
  chmod 755 palette-edge
  ```

> 3. Use the following command to move the **palette-edge** binary to the **/bin/** directory. You will need elevated privileges to perform this operation. This step will allow you to issue the `palette-edge` command from any directory in your development environment.

Reviewer: ```suggestion
3. Use the following command to move the palette-edge binary to the **/usr/local/bin** directory so that the binary is available in your system $PATH.  This will allow you to issue the `palette-edge` command from any directory in your development environment.
```

Use `/usr/local/bin` instead of `/bin`. `/bin` is often a protected file folder so removing binaries is often prohibited. This is the case for Mac environments, the preferred approach is for `/usr/local/bin` as the binary is scoped to the current users. Now, if you know for sure that the binary needs to be globally scoped then use `/bin` but do so with care.

<br /> 

  ```bash

>   sudo mv palette-edge /bin/

Reviewer: ```suggestion
  sudo mv palette-edge /usr/local/bin
```

  ```

> 4. Verify the installation of the Palette Edge CLI by issuing the following command. The output will display information about the installed version of the Palette Edge CLI. 

Reviewer: ```suggestion
4. Verify the installation of the Palette Edge CLI by issuing the following command. The output will display information about the currently supported OS and Kubernetes distributions. 
```
You are not using the version command? Is that intentional?

  | OS Flavor     | Description        | Base Image URI                    |
  | opensuse-leap | Opensuse Leap 15.4 | quay.io/kairos/core-opensuse-leap |
  | ubuntu-20     | Ubuntu 20.4 LTS    | quay.io/kairos/core-ubuntu-20-lts |
  | ubuntu-22     | Ubuntu 22.4 LTS    | quay.io/kairos/core-ubuntu-22-lts |
  └────────────────────────────────────────────────────────────────────────┘
  ┌─────────────────────────────────────────────────────────────────────────────────────────────┐
  | K8S Flavor | Description        | Supported Versions                                        |
  | k3s        | Rancher K3s        | 1.25.2-k3s1,1.24.6-k3s1,1.23.12-k3s1,1.22.15-k3s1         |
  | kubeadm    | Kubernetes kubeadm | 1.25.2,1.24.6,1.23.12,1.22.15                             |
  | rke2       | Rancher RK2        | 1.25.2-rke2r1,1.24.6-rke2r1,1.23.12-rke2r1,1.22.15-rke2r1 |
  └─────────────────────────────────────────────────────────────────────────────────────────────┘
  ┌─────────────────────────────────┐
  | Component             | Version |
  | Spectro Agent Version | v3.4.3  |
  | Kairos Version        | v2.0.3  |
  └─────────────────────────────────┘
  ```
<br />

> 5. The next few steps will guide you to create a cluster profile referencing the provider image you must have already pushed to an image registry. Check the **Prerequisites** section if you do not have a provider image in an image registry. 
> 
>   Open a web browser and log in to [Palette](https://console.spectrocloud.com). Ensure you are in the **Default** project scope before creating a cluster profile. 

Reviewer: ```suggestion
5. Open a web browser and log in to [Palette](https://console.spectrocloud.com). Ensure you are in the **Default** project scope before creating a cluster profile. 
```

This is not a tutorial. Go to the next steps. There is no need for this verbosity when you provided an overview in the beginning. How-tos are concise and to the point.

6. Navigate to the left **Main Menu** and select **Profiles**. Click on the **Add Cluster Profile** button, and fill out the required basic information fields to create a cluster profile for Edge. 

7. Add the following OS layer in the **Profile Layers** section.

  |**Pack Type**|**Registry**|**Pack Name**|**Pack Version**| 
  |---|---|---|---|
  |OS|Public Repo|BYOS Edge OS|`1.0.0`|

> 8. Replace the OS layer manifest with the following custom manifest so that the cluster profile can pull the provider image from the ttl.sh image registry. This manifest is the output CanvOS returned after building the Edge artifacts in the [Build Edge Artifacts](/clusters/edge/edgeforge-workflow/palette-canvos) workflow.

Reviewer: If the Build Edge Artifacts guide is a prerequisite, why not reference the cluster profile they created when completing the guide? That would reduce several steps from this guide. The only thing we need is the cluster profile ID. 

We can essentially remove steps 7 - 12.

  </WarningBox>
  <br />

9. Add the following Kubernetes layer to your cluster profile. Select the K3s version 1.25.x because earlier in this how-to guide, you pushed a provider image compatible with K3s v1.25.2 to the ttl.sh image registry. 

  |**Pack Type**|**Registry**|**Pack Name**|**Pack Version**| 
  |---|---|---|---|
  |Kubernetes|Public Repo|Palette Optimized K3s|`1.25.x`|

10. Add the network layer to your cluster profile, and choose a Container Network Interface (CNI) pack that best fits your needs, such as Calico, Flannel, Cilium, or Custom CNI. For example, you can add the following network layer. This step completes the core infrastructure layers in the cluster profile.  

  |**Pack Type**|**Registry**|**Pack Name**|**Pack Version**| 
  |---|---|---|---|
  |Network|Public Repo|Calico|`3.25.x`|
    
  
11. Add add-on layers and manifests to your cluster profile per your requirements. 

12. If there are no errors or compatibility issues, Palette displays the newly created complete cluster profile for review. Verify the layers you added, and finish creating the cluster profile. 

> 13. The remaining steps will guide you to create an offline content bundle using the Palette Edge CLI and build the Edge installer ISO. 

Reviewer: Step 13 should be a prereq, at least the steps on how to create the API key. If anything, we need to improve the steps [here](https://docs.spectrocloud.com/user-management/user-authentication/#creatinganapikeyasatenantadmin) and just link to this section.

  To use Palette Edge CLI, you need a Spectro Cloud API key to authenticate and interact with the Palette API endpoint. To add a new API key, log in to Palette, click on the user **User Menu** at the top right, and select **My API Keys**, as displayed in the screenshot below. 

  Fill in the required fields, such as the API key name and expiration date, and confirm your changes. Copy the key value to your clipboard to use in the next step.

  ![Screenshot of generating an API key in Palette.](/tutorials/deploy-pack/registries-and-packs_deploy-pack_generate-api-key.png )

  

> 14. Set the API key by issuing the following command. Replace the `[USE-YOUR-API-KEY_HERE]` placeholder with your API key. The Palette Edge CLI will use this API key to authenticate itself when interacting with the Spectro Cloud API. Once authenticated, the Palette Edge CLI can access and manage resources in your Spectro Cloud account.

Reviewer: ```suggestion
14. Set the API key as an environment variable by issuing the following command. Replace the `[USE-YOUR-API-KEY_HERE]` placeholder with your API key. The Palette Edge CLI will use this API key to authenticate itself when interacting with the Spectro Cloud API. Once authenticated, the Palette Edge CLI can interact with your Palette account.
```
The Edge CLI does not manage Edge resources.

<br />

  ```bash
  export API_KEY=[USE-YOUR-API-KEY_HERE]
  ```

> 15. Copy the cluster profile and project IDs from Palette to use in the next step. To locate these IDs, switch to [Palette](https://console.spectrocloud.com). Navigate to the left **Main Menu** and select **Profiles**. Click on the specific cluster profile to view its details. The cluster profile and project IDs are in the cluster details page URL. The cluster details page URL follows the `[Palette-URL]/projects/[PROJECT-ID]/profiles/cluster/[PROFILE-ID]` syntax. 

Reviewer: Break these steps up. You are overloading several steps into one.  Also, let's show the user how to find the projectID the right way (see screenshot). The Cluster Profile ID is something I think the UX team is improving in the next release. 

![CleanShot 2023-07-24 at 15 30 01](https://github.com/spectrocloud/librarium/assets/29551334/e9acd2d7-3f41-46d9-8f88-145839b645c1)

  For example, the screenshot below highlights the cluster profile and project IDs in a cluster details page URL. 

  ![A screenshot highlighting the cluster profile and project IDs in the URL of the cluster details page.](/clusters_edge_edge-forge-workflow_build-images_build-artifacts_url.png)

> 16. Set the profile ID using the following command. Replace the `[USE-YOUR-PROFILE-ID_HERE]` placeholder with your cluster profile ID. This command sets the profile ID, which identifies the specific cluster profile that you want to use for creating the offline content bundle. 

Reviewer: ```suggestion
16. Set the cluster profile ID using the following command. Replace the `[USE-YOUR-PROFILE-ID_HERE]` placeholder with your cluster profile ID. The Palette Edge CLI uses the cluster profile ID to reference the correct cluster profile and download all of its software dependencies. 
```

<br />

  ```bash
  export PROFILE_ID=[USE-YOUR-PROFILE-ID_HERE]
  ```

> 17. Set the project ID by issuing the following command. Replace the `[USE-YOUR-PROJECT-ID_HERE]` placeholder with your cluster profile ID.
> The offline content bundle will be associated with the project ID you will use in the following command. 

Reviewer: ```suggestion
17. Set the project ID by issuing the following command. Replace the `[USE-YOUR-PROJECT-ID_HERE]` placeholder with your project ID.
```

<br />

  ```bash
  export PROJECT_ID=[USE-YOUR-PROJECT-ID_HERE]
  ```

> 18. Build the offline content bundle using the following command. The content bundle will have the configuration and settings for the Edge cluster, including the operating system image, Kubernetes distribution, any additional packs, and manifest files specified in your cluster profile you wish to deploy to the Edge cluster.
> 
>   The command below uses the palette-edge tool to build the offline content bundle. The `--api-key`, `--project-id`, and `--cluster-profile-ids` options specify the necessary parameters for the build process. The `--palette-endpoint` option specifies the endpoint URL for the Spectro Cloud API. The `--outfile` option specifies the name of the output file for the content bundle.

Reviewer: ```suggestion
18.  The `build` command creates the content bundle However, the command requires a set of parameters to start. The `--api-key`, `--project-id`, `--cluster-profile-ids`, and  `--palette-endpoint` are required flags.  Use the `--outfile` option to specify the name of the output file for the content bundle. Issue the following command to start the content bundle build process.
```

We can simplify this. Remember, you mentioned this in the intro (see below) so you don't have to repeat yourself here. You emphasize this in the conclusion but it's necessary here. 

_A content bundle is an archive that includes the operating system image, Kubernetes distribution, additional layers, and packs and manifest files specified in your cluster profile that you want to deploy to the Edge cluster. A content bundle provides several benefits, such as:_

<br />

  ```bash
  palette-edge build --api-key $API_KEY \
  --project-id $PROJECT_ID \
  --cluster-profile-ids $PROFILE_ID \
  --palette-endpoint api.spectrocloud.com \
  --outfile content
  ```

19. Use the command below to list all files in the current directory, including the newly built content bundle. The content bundle will have the following naming convention, `content-[randon-string]`, for example, **content-8e61a9e5**. The command below will verify that you successfully created the content bundle with the specified name.
<br />

  ```bash
  ls -al
  ```

20. List the details of the content bundle using the following command. Replace the `content-[randon-string]` with the content bundle in your current directory. The command below will provide information such as file size, permissions, and modification date.
<br />

  ```bash
  ls -l content-[randon-string]
  ```

> 21. Issue the following command to execute the **earthly.sh** script with elevated privileges. The `+iso` option specifies the build target. This command will generate an ISO image from the content bundle and other configurations you have specified in the **.arg** and **user-data** files. 

Reviewer: Should we mention that it takes a while 😅 . You used this sentence in the Edge artifacts guide:
_This command may take up to 15-20 minutes to finish depending on the resources of the host machine_.

<br />

  ```bash
  sudo ./earthly.sh +iso
  ```

> 22. List the Edge installer ISO image and checksum by issuing the following command from the **CanvOS/** directory.

Reviewer: ```suggestion
22. Once the build process is complete, list the Edge installer ISO image and checksum by issuing the following command from the **CanvOS/** directory.
```

<br />

  ```shell
  ls build/
  ```

  ```shell hideClipboard
  # Output
  palette-edge-installer.iso      
  palette-edge-installer.iso.sha256
  ```

> # Validate

Reviewer: For the Validate section you are in the same spot as you were in the Edge artifacts guide. You can use the same language you used there but you can explain that in order to test, the user must deploy an Edge cluster.

To validate the Edge installer ISO image, you can follow these steps:
<br />

1. Create a bootable USB flash drive using any third-party software.

2. Select a physical or virtual host machine to emulate as an edge device. Enable (Dynamic Host Configuration Protocol) DHCP on the edge device before proceeding with the installation process. Enabling DHCP is necessary for the device to obtain an IP address automatically from the network. 

> 3. Use the software to flash a bare host machine with a bootable USB drive. Most software that creates a bootable USB drive will automatically validate the ISO image during the installation process. 

Reviewer: Add a Next Steps that provides a summary and guides users to check out the following resources. 

[Deploy an Edge Cluster on VMware](https://deploy-preview-1400--docs-spectrocloud.netlify.app/clusters/edge/site-deployment/deploy-cluster)
[Prepare Edge Host for Installation](https://deploy-preview-1400--docs-spectrocloud.netlify.app/clusters/edge/site-deployment/stage)
